Source/CarthageKit/Project.swift

```diff
@@ -79,6 +79,24 @@ extension ProjectEvent: Equatable {
 	}
 }
 
+/// Use these queues to ensure there is only one write operation in a specific local
+/// repository directory at a time.
+internal struct RepositoryOperationQueues {
+	private static var queues: Atomic<[URL: SerialProducerQueue]> = Atomic([:])
+
+	internal static func queue(forLocalRepositoryURL url: URL) -> SerialProducerQueue {
+		return queues.modify { queues -> SerialProducerQueue in
+			if let queue = queues[url] {
+				return queue
+			} else {
+				let newQueue = SerialProducerQueue(name: "org.carthage.Constants.Project.cloneOrFetchQueue-\(url.absoluteString)")
+				queues[url] = newQueue
+				return newQueue
+			}
+		}
+	}
+}
```

Comment:
This global state makes me very uncomfortable. I don't think it's a good idea to introduce a `static var` with this. It leads to code that's hard to follow and modify.

We should use another method to enforce exclusivity. I think it makes sense to do something like this:

1. Fetch all the submodules from the repository cache, in parallel but with limited concurrency.
2. Perform the actual checkouts, passing in a queue or doing them one at a time

Reply:
I see there are already a lot of queues, but they are instance properties of a Project. So I just need to move the global variable to the Project class?